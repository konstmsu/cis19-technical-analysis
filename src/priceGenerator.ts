import { linspace } from "./math";
import { sin } from "mathjs";

export function randomSeq(count: number) {
  const result = [];

  for (let i = 0; i < count; i++) {
    result.push(Math.random() * 2 - 1.0);
  }

  return result;
}

export function linearSeq(start: number, stop: number) {
  return (count: number) => linspace(start, stop, count);
}

export function sinSeq(start: number, stop: number) {
  return (count: number): number[] => {
    return linspace(start, stop, count).map(x => sin(x));
  };
}
