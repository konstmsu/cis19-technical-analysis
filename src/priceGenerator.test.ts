import { max, min, std, pi, sqrt } from "mathjs";
import { randomSeq, sinSeq } from "./priceGenerator";

describe("price generator", () => {
  it("generate random numbers", () => {
    const seq = randomSeq(100);

    expect(seq).toHaveLength(100);
    expect(0).toBeWithin(min(...seq), max(...seq));
    expect(std(seq)).toBeGreaterThan(0.5);
  });

  it("generate sin wave", () => {
    const seq = sinSeq(0, pi)(13);
    const rounded = (values: number[]) => values.map(v => v.toFixed(5));
    expect(rounded(seq)).toEqual(
      rounded([
        0,
        0.25882,
        1 / 2,
        1 / sqrt(2),
        sqrt(3) / 2,
        0.96593,
        1,
        0.96593,
        sqrt(3) / 2,
        1 / sqrt(2),
        1 / 2,
        0.25882,
        0
      ])
    );
  });

  it("generate longer sin wave", () => {
    const seq = sinSeq(0, pi * 4)(25);
    const rounded = (values: number[]) => values.map(v => v.toFixed(5));
    expect(rounded(seq)).toMatchSnapshot();
  });
});
