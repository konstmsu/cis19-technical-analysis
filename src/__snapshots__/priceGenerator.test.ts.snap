// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`price generator generate longer sin wave 1`] = `
Array [
  "0.00000",
  "0.50000",
  "0.86603",
  "1.00000",
  "0.86603",
  "0.50000",
  "0.00000",
  "-0.50000",
  "-0.86603",
  "-1.00000",
  "-0.86603",
  "-0.50000",
  "-0.00000",
  "0.50000",
  "0.86603",
  "1.00000",
  "0.86603",
  "0.50000",
  "0.00000",
  "-0.50000",
  "-0.86603",
  "-1.00000",
  "-0.86603",
  "-0.50000",
  "-0.00000",
]
`;
